### ISSUE TITLE
Error with async: "Parsing error: Unexpected token =>"

### ISSUE THREAD

**Issue by huarse** [eslint/eslint#8126]
* **ESLint Version:** 3.16.0
* **Node Version:** 6.9.1
* **npm Version:** 3.10.8

* **parser:** default
* **configuration**

```js
...
"extends": ["eslint:recommended"],
"parserOptions": {
    "sourceType": "module",
    "ecmaVersion": 7,
    "ecmaFeatures": {
        "jsx": true,
        "experimentalObjectRestSpread": true
    }
},
...
```

**source code**

```js
!(async () => {
    try {
        const result = await Promise.resolve('Hello');
        console.log('>>>>>', result);
    } catch (error) {
        console.error('Error in demo', error);
    }
})();
```

**eslint check result**

```
error  Parsing error: Unexpected token =>
```

**Comment by not-an-aardvark** [comment]
Thanks for the report. Async functions are an ES2017 (or "ES8") feature, but you have `ecmaVersion: 7` in your config. To fix this, you should use `ecmaVersion: 8` or `ecmaVersion: 2017` in your config.

**Comment by huarse** [comment]
THX @not-an-aardvark, it works. I always thought that 'async' is ES7 features

**Comment by anwarhamr** [comment]
I'm still having this issue in Atom.  I've tried both values `ecmaVersion: 8` and  `ecmaVersion: 2017` in my config.  I'm using Atom editor w/ the linter packages: linter 2.2.0, linter-eslint 8.3.2, linter-ui-default 1.6.10.  
Here is the code failing:
```
export default class myclass{
  async handleDiscoverPeripheral(peripheral) { ...}
}
```

my .eslintrc.json:
```
{
    "env": {
        "browser": true,
        "node": true,
        "es6": true
    },
    "extends": ["eslint:recommended", "plugin:react/recommended"],
    "parserOptions": {
        "ecmaFeatures": {
          "ecmaVersion": 8,
            "experimentalObjectRestSpread": true,
            "jsx": true
        },
        "sourceType": "module"
    },
    "plugins": [
        "react"
    ],
    "rules": {
        "no-console": 0,
        "disallowMultipleVarDecl": 0,
        "maximumLineLength": 0,
        "indent": [
            "error",
            2
        ],
        "linebreak-style": [
            "error",
            "unix"
        ],
        "quotes": [
            "error",
            "single"
        ],
        "semi": [
            "error",
            "always"
        ]
    }
}
```

**Comment by not-an-aardvark** [comment]
`ecmaVersion` should not be in `ecmaFeatures`.

```diff
{
    "env": {
        "browser": true,
        "node": true,
        "es6": true
    },
    "extends": ["eslint:recommended", "plugin:react/recommended"],
    "parserOptions": {
+       "ecmaVersion": 8,
        "ecmaFeatures": {
-           "ecmaVersion": 8,
            "experimentalObjectRestSpread": true,
            "jsx": true
        },
        "sourceType": "module"
    },
    "plugins": [
        "react"
    ],
    "rules": {
        "no-console": 0,
        "disallowMultipleVarDecl": 0,
        "maximumLineLength": 0,
        "indent": [
            "error",
            2
        ],
        "linebreak-style": [
            "error",
            "unix"
        ],
        "quotes": [
            "error",
            "single"
        ],
        "semi": [
            "error",
            "always"
        ]
    }
}
```

**Comment by henryzp** [comment]
@huarse ，when i use ecmaVersion: 8

![image](https://user-images.githubusercontent.com/3849702/32087372-f802d572-baa0-11e7-859e-253528b99d5a.png)

@not-an-aardvark , can you help me ？

**Comment by Aladdin-ADD** [comment]
@henryzp it might be you are using a very earlier version ESLint (does not supporting ES8), upgrade to latest should work~

**Comment by henryzp** [comment]
@Aladdin-ADD 

I use vscode:

![image](https://user-images.githubusercontent.com/3849702/32088689-e61f8a94-baaa-11e7-97cf-f44aab1cf6bd.png)

I guess its version is not the latest。。。

**Comment by Aladdin-ADD** [comment]
well, you can upgrade and try. if still got some problems, please feel free to leave a comment.

just a friendly reminder: it seems to be vscode-eslint version, not eslint. you can use `eslint -v`(global installed) or `./node_modules/.bin/eslint -v`(local-installed) to see eslint verson.

**Comment by kstiopin** [comment]
Have the same issue and have tried all possible solutions mentioned here and in similar topics.
So this is the error:
`Syntax error: C:/projects/project1/src/containers/OfferOverview.js: Unexpected token (84:40)
  83 |
> 84 |   initialGetProduct = async (productId) => {
       |                                                                  ^
  85 |     await dispatch(resetOfferStateAction());
  86 |     dispatch(getProductAction(productId));
  87 |     this.selectProduct(productId);`

This is the Component:
`export class OfferOverview extends Component {
  componentWillMount() {
    const { location } = this.props;
    const productId = location.query && location.query.product_id;
    if (productId) {
      this.initialGetProduct(productId);
    }
  }

  initialGetProduct = async (productId) => {
    await dispatch(resetOfferStateAction());
    dispatch(getProductAction(productId));
    this.selectProduct(productId);
  }

  render() { ... }
}`

this is my .babelrc file:
`{
  "presets": [
    "es2017",
    "react",
    "stage-0"
  ],
  "plugins": [
    "react-hot-loader/babel",
    ["import", { "libraryName": "antd", "style": "css" }]
  ]
}
`
eslint:
`module.exports = {
  root: true,

  parser: 'babel-eslint',

  // import plugin is termporarily disabled, scroll below to see why
  plugins: [/*'import', */'flowtype', 'jsx-a11y', 'react'],

  env: {
    browser: true,
    commonjs: true,
    node: true
  },

  parserOptions: {
    ecmaVersion: 2017,
    sourceType: 'module',
    ecmaFeatures: {
      jsx: true,
      generators: true,
      experimentalObjectRestSpread: true
    }
  },

  settings: {
    'import/ignore': [
      'node_modules',
      '\\.(json|css|jpg|png|gif|eot|svg|ttf|woff|woff2|mp4|webm)$',
    ],
    'import/extensions': ['.js'],
    'import/resolver': {
      node: {
        extensions: ['.js', '.json']
      }
    }
  },

  rules: { ... }
};`

and package.json dependencies: 
`"devDependencies": {
    "babel-cli": "^6.26.0",
    "babel-core": "^6.13.2",
    "babel-eslint": "^6.1.2",
    "babel-plugin-transform-runtime": "^6.12.0",
    "babel-polyfill": "^6.26.0",
    "babel-preset-es2015": "^6.13.2",
    "babel-preset-es2016": "^6.24.1",
    "babel-preset-es2017": "^6.24.1",
    "babel-preset-react": "^6.11.1",
    "babel-preset-stage-2": "^6.13.0",
    "coveralls": "^2.11.12",
    "eslint": "^3.2.2",
    "eslint-config-airbnb": "^10.0.0",
    "eslint-plugin-react": "^6.0.0",
    "ignore-styles": "^4.0.0",
    "istanbul": "^1.0.0-alpha.2",
    "mocha": "^3.0.2",
    "nock": "^8.0.0",
    "react-addons-test-utils": "^15.3.0",
    "react-scripts": "0.2.1",
    "redux-mock-store": "^1.1.2",
    "redux-saga-devtools": "^0.1.2"
  },
  "dependencies": {
    "antd": "^2.12.3",
    "babel-polyfill": "^6.26.0",
    "enzyme": "^2.4.1",
    "eslint-plugin-flowtype": "^2.39.1",
    "eslint-plugin-jsx-a11y": "^6.0.2",
    "expect": "latest",
    "isomorphic-fetch": "^2.2.1",
    "jsdom": "^9.9.1",
    "lodash": "^4.17.4",
    "nuka-carousel": "^2.3.0",
    "pushstate-server": "latest",
    "react": "^15.2.1",
    "react-dom": "^15.2.1",
    "react-redux": "^4.4.5",
    "react-router": "^2.6.0",
    "react-router-redux": "^4.0.8",
    "react-sidebar": "^2.3.2",
    "redux": "^3.6.0",
    "redux-logger": "3.0.1",
    "redux-saga": "^0.14.3"
  },`

Would appreciate any hint what can be wrong?

**Comment by platinumazure** [comment]
@kstiopin This seems to be a babel-eslint issue. Could you please open an issue there? Thanks!
